examples/model_checking.py

```diff
@@ -35,115 +36,205 @@ def __init__(self, variables, init, trans):
+        return False
+        
+class BMCInduction(object):
+    
+    def __init__(self, system):
+        self.system = system
+    
+    def get_subs(self, i):
+        """Builds a map from x to x@i and from x' to x@(i+1), for all x in system."""
+        subs_i = {}
+        for v in self.system.variables:
+            subs_i[v] = at_time(v, i)
+            subs_i[next_var(v)] = at_time(v, i+1)
+        return subs_i
+
+
+    def get_unrolling(self, k):
+        """Unrolling of the transition relation from 0 to k:
+
+        E.g. T(0,1) & T(1,2) & ... & T(k-1,k)
+        """
+        res = []
+        for i in xrange(k+1):
+            subs_i = self.get_subs(i)
+            res.append(self.system.trans.substitute(subs_i))
+        return And(res)
+
+
+    def get_simple_path(self, k):
+        """Simple path constraint for k-induction:
+        each time encodes a different state
+        """
+        res = []
+        for i in xrange(k+1):
+            subs_i = self.get_subs(i)
+            for j in xrange(i+1, k+1):
+                state = []
+                subs_j = self.get_subs(j)
+                for v in self.system.variables:
+                    v_i = v.substitute(subs_i)
+                    v_j = v.substitute(subs_j)
+                    state.append(Not(EqualsOrIff(v_i, v_j)))
+                res.append(Or(state))
+        return And(res)
+
+
+    def get_k_hypothesis(self, prop, k):
+        """Hypothesis for k-induction: each state up to k-1 fulfills the property"""
+        res = []
+        for i in xrange(k):
+            subs_i = self.get_subs(i)
+            res.append(prop.substitute(subs_i))
+        return And(res)
+
+
+    def get_bmc(self, prop, k):
+        """Returns the BMC encoding at step k"""
+        init_0 = self.system.init.substitute(self.get_subs(0))
+        prop_k = prop.substitute(self.get_subs(k))
+        return And(self.get_unrolling(k), init_0, Not(prop_k))
+
+    def get_k_induction(self, prop, k):
+        """Returns the K-Induction encoding at step K"""
+        subs_k = self.get_subs(k)
+        prop_k = prop.substitute(subs_k)
+        return And(self.get_unrolling(k),
+                   self.get_k_hypothesis(prop, k),
+                   self.get_simple_path(k),
+                   Not(prop_k))
+
+    def check_property(self, prop):
+        """Interleaves BMC and K-Ind to verify the property."""
+        print("Checking property %s..." % prop)
+        for b in xrange(100):
+            f = self.get_bmc(prop, b)
+            print("   [BMC]    Checking bound %d..." % (b+1))
+            if is_sat(f):
+                print("--> Bug found at step %d" % (b+1))
+                return
+
+            f = self.get_k_induction(prop, b)
+            print("   [K-IND]  Checking bound %d..." % (b+1))
+            if is_unsat(f):
+                print("--> The system is safe!")
+                return
+
+
+def counter(bit_count):
+    """Counter example with n bits and reset signal."""
```

Comment:
This example is pre-BV. It makes sense to convert it to BVs. Both approaches should still work.